proxy/http/HttpSM.cc

```diff
@@ -3161,15 +3172,25 @@ HttpSM::tunnel_handler_ua(int event, HttpTunnelConsumer *c)
     set_ua_abort(HttpTransact::ABORTED, event);
 
     if (is_bg_fill_necessary(c)) {
+      p     = c->producer;
+      selfc = p->self_consumer;
+      if (selfc && selfc->producer->vc_type == HT_HTTP_SERVER) {
+        p = selfc->producer;
+      }
+
       DebugSM("http", "[%" PRId64 "] Initiating background fill", sm_id);
-      background_fill = BACKGROUND_FILL_STARTED;
+      // check whether to finish the reading.
+      background_fill = p->read_success ? BACKGROUND_FILL_COMPLETED : BACKGROUND_FILL_STARTED;
       HTTP_INCREMENT_DYN_STAT(http_background_fill_current_count_stat);
```

Comment:
You forget to decrement the http_background_fill_current_count_stat if BACKGROUND_FILL_COMPLETED